src/game/traps/movable_block.c

```diff
@@ -9,6 +9,40 @@
 #include "game/lara.h"
 #include "game/sound.h"
 #include "global/vars.h"
+#include "src/game/collide.h"
+#include "src/game/sphere.h"
+
+static bool TestDoorCoveringBlock(ITEM_INFO *lara_item, COLL_INFO *coll);
+
+static bool TestDoorCoveringBlock(ITEM_INFO *lara_item, COLL_INFO *coll)
+{
+    // OG fix: stop pushing blocks through doors
+    int32_t max_dist = SQUARE((WALL_L * 2) >> 8);
+    for (int item_num = 0; item_num < g_LevelItemCount; item_num++) {
+        ITEM_INFO *item = &g_Items[item_num];
+        int32_t dx = (item->pos.x - lara_item->pos.x) >> 8;
+        int32_t dy = (item->pos.y - lara_item->pos.y) >> 8;
+        int32_t dz = (item->pos.z - lara_item->pos.z) >> 8;
+        int32_t dist = SQUARE(dx) + SQUARE(dy) + SQUARE(dz);
+
+        if (dist > max_dist) {
+            continue;
+        }
+
+        if ((item->object_number < O_DOOR_TYPE1
+             || item->object_number > O_DOOR_TYPE8)) {
+            continue;
+        }
+
+        if (TestBoundsCollide(item, lara_item, coll->radius)) {
+            return true;
+        }
+        if (TestCollision(item, lara_item)) {
```

Comment:
Don't we need to have both of these conditions satisfied at once in order to detect a collision?
In that case I'd say

if (!Test…()) continue;
if (!Test…()) continue;
return true;

Reply:
I don't think both need to be satisfied at once? Here's DoorCollision (which is for a single door):
```
void DoorCollision(int16_t item_num, ITEM_INFO *lara_item, COLL_INFO *coll)
{
    ITEM_INFO *item = &g_Items[item_num];

    if (!TestBoundsCollide(item, lara_item, coll->radius)) {
        return;
    }
    if (!TestCollision(item, lara_item)) {
        return;
    }

    if (coll->enable_baddie_push) {
        if (item->current_anim_state != item->goal_anim_state) {
            ItemPushLara(item, lara_item, coll, coll->enable_spaz, 1);
        } else {
            ItemPushLara(item, lara_item, coll, 0, 1);
        }
    }
}
```

Reply:
DoorCollision does not push Lara if TestBoundsCollide fails or TestCollision fails. Applying De Morgan's law to flip this expression, we get: DoorCollision pushes Lara if TestBoundsCollide succeeds and TestCollision succeeds.

They need to be both satisfied.

Reply:
Yes true. Still not used to the backwards if checking lol. So what about:
```
static bool TestDoorCoveringBlock(ITEM_INFO *lara_item, COLL_INFO *coll)
{
    // OG fix: stop pushing blocks through doors
    int32_t max_dist = SQUARE((WALL_L * 2) >> 8);
    for (int item_num = 0; item_num < g_LevelItemCount; item_num++) {
        ITEM_INFO *item = &g_Items[item_num];
        int32_t dx = (item->pos.x - lara_item->pos.x) >> 8;
        int32_t dy = (item->pos.y - lara_item->pos.y) >> 8;
        int32_t dz = (item->pos.z - lara_item->pos.z) >> 8;
        int32_t dist = SQUARE(dx) + SQUARE(dy) + SQUARE(dz);

        if (dist > max_dist) {
            continue;
        }

        if ((item->object_number < O_DOOR_TYPE1
             || item->object_number > O_DOOR_TYPE8)) {
            continue;
        }

        if (TestBoundsCollide(item, lara_item, coll->radius) && TestCollision(item, lara_item)) {
            return true;
        }
    }
    return false;
}
```

I'm not sure what you mean about the loop ending in a return? In this method, it checks all the doors nearby. If Lara collides with one, return true. Otherwise if all doors get checked and there's no collision, return false.